File: 프로그래머스/lv2/76502. 괄호 회전하기/괄호 회전하기.py
from collections import deque
def solution(s):
    answer = 0
    n = len(s)
    key = {'}':'{', ']':'[', ')':'('}
    for i in range(n):
        stack = deque()
        done = True
        for j in range(n):
            index = i + j
            if index >= n:
                index -= n
            if s[index] == '[' or s[index] == '(' or s[index] == '{':
                stack.append(s[index])
            else:
                if len(stack) > 0 and key[s[index]] == stack[-1]:
                    stack.pop()
                else:
                    done = False
                    break
        if done and len(stack) == 0:
            answer += 1
    return answer

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;

class Solution {
    int answer, n;
    String s;
    public int solution(String s) {
        init(s);
        solve();
        return answer;
    }
    
    void init(String s) {
        this.answer = 0;
        this.s = s;
        this.n = s.length();
    }
    
    void solve() {
        for (int i = 0; i < n; i++) {
            if (isRight()) {
                answer += 1;
            } 
            s = s.substring(1) + s.substring(0,1);
        }
    }
    
    boolean isRight() {
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < n; i++) {
            char c = s.charAt(i);
            if (c == '(' || c == '[' || c == '{') {
                stack.add(c);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }
                Character removed = stack.pop();
                if (c == ')' && removed == '(') {
                    continue;
                }
                if (c == ']' && removed == '[') {
                    continue;
                }
                if (c == '}' && removed == '{') {
                    continue;
                }
                return false;
            }
            
        }
        if (!stack.isEmpty()) {
            return false;
        }
        return true;
    }
}